qutebrowser/misc/tabsfinder.py

```diff
@@ -0,0 +1,92 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+import re
+from collections import OrderedDict, namedtuple
+
+from qutebrowser.utils import objreg
+
+Tab = namedtuple('Tab', 'url title index')
+
+
+class TabsFinder:
```

Comment:
Here's my 2 cents on this, though others may have a different philosophy :grin: 

I don't love the use of a class here, because:

1. It computes several different data structures, only one of which is necessary each place it is used.
2. It does non-trivial work in a constructor, which can be surprising
3. I'm not a fan of noun-ifying verbs. That is, don't use a class when a function would do (this probably qualifies as bikeshedding, though)

I would be more keen on one or more module-level functions:

1. `all_tabs`, which yields* successive (index, url, title) tuples.
2. `all_tabs_by_window` which groups `all_tabs` by `win_id` _or_ just omit this and do the grouping in the completion code

`index_for_url` probably  doesn't need a helper, it should be as simple as `idx = next((t for t in all_tabs()) if t.url == url), None)`

\* I think it makes sense as a generator since the evaluation will be transformed (for completion) or stopped early (for finding a matching tab). That being said, this isn't likely to be performance critical, so if using generators for some reason complicates things then ignore me :grin:

Reply:
Maybe also consider returning (win_id, index, url, title) tuples, as that feels more "pure-data"-y, and only form strings when needed for display. Ignore this if it makes things more complicated though